src/mpsc/async_impl.rs

```diff
@@ -281,6 +281,161 @@ feature! {
         }
     }
 
+    feature! {
+        #![feature = "stream"]
+        use futures::sink::Sink;
+        use std::mem;
+
+        /// PollSender internal state.
+        enum PollSenderState<'a, T, R> {
+            Idle(Sender<T, R>),
+            Acquiring(Pin<Box<SendRefFuture<'a ,T, R>>>),
+            ReadyToSend(SendRef<'a, T>),
+            Closed,
+        }
+
+        /// A wrapper around mpsc::Sender that can be polled.
+        pub struct PollSender<'a, T, R> {
+            sender: Option<Sender<T, R>>,
+            state: PollSenderState<'a, T, R>,
+        }
+
+        impl<'a, T, R> PollSender<'a, T, R> where R: 'a + Recycle<T>, T: 'a {
+            /// Creates a new `PollSender`.
+            pub fn new(sender: Sender<T, R>) -> PollSender<'a, T, R> {
+                PollSender {
+                    sender: Some(sender.clone()),
+                    state: PollSenderState::Idle(sender),
+                }
+            }
+
+            /// Attempts to reserve a slot in the channel.
+            pub fn poll_reserve(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Closed>> {
+                loop {
+                    let (result, next_state) = match mem::replace(&mut self.state, PollSenderState::Closed) {
+                        PollSenderState::Idle(sender) => {
+                            let send_ref = Box::pin(SendRefFuture {
+                                core: &sender.inner.core,
+                                slots: sender.inner.slots.as_ref(),
+                                recycle: &sender.inner.recycle,
+                                state: State::Start,
+                                waiter: queue::Waiter::new(),
+                            });
```

Comment:
hmm, it's unfortunate that every time a `poll_sender` sends a new message, we'll allocate a new `Box` which is dropped when that message is sent. it would be preferable if there was a single heap allocation that lives for the entire lifetime of the sender, instead. 

`tokio_util`'s `PollSender` uses a [`ReusableBoxFuture` type](https://docs.rs/tokio-util/latest/src/tokio_util/sync/mpsc.rs.html#47) for this. I would be open to adding an optional `tokio-util` dependency that's only enabled when the pollable sender implementation is used, if it allowed us to avoid the allocation that occurs on every `poll_reserve` call...

Reply:
Ok. Sure! I saw `ReusableBoxFuture` but have been afraid of adding an external dependent.